File: NN/NN.cpp
// NN.cpp : Этот файл содержит функцию "main". Здесь начинается и заканчивается выполнение программы.
//

#include <iostream>
#include "Neuron.h"
#include "Layer.h"
#include "Model.h"

int main()
{
    sigmoida sigma;
    int a = 10;
    input_layer l(3);
    output_layer out(1,&sigma);

    nn mod;

    mod.add_layers(&l);
    mod.add_layers(&out);

    mod.compile();

    mod.layers[1]->neurons[0]->w[0] = -0.16595599;
    mod.layers[1]->neurons[0]->w[1] = 0.44064899;
    mod.layers[1]->neurons[0]->w[2] = -0.99977125;

    /*mod.layers[1]->neurons[0]->w[0] = 9.67299303;
    mod.layers[1]->neurons[0]->w[1] = -0.2078435;
    mod.layers[1]->neurons[0]->w[2] = -4.62963669;*/

    vector<double> x1 = { 0, 0, 1 };
    vector<double> x2 = { 1, 1, 1 };
    vector<double> x3 = { 1, 0, 1 };
    vector<double> x4 = { 0, 1, 1 };
    vector<double> x5 = { 1, 0, 0 };

    vector<double> y1 = {0};
    vector<double> y2 = {1};
    vector<double> y3 = {1};
    vector<double> y4 = {0};


    vector<double> grad = { 0 };

    vector<double> predict;

    for (int i = 0; i < 10000; i++)
    {
        std::cout << i << "\n";
        predict = mod.forward(x1);
        grad[0] = 2 * ( predict[0]-y1[0]);
        mod.backward(grad);

        predict = mod.forward(x2);
        grad[0] = 2 * (predict[0] - y2[0]);
        mod.backward(grad);

        predict = mod.forward(x3);
        grad[0] = 2 * (predict[0] - y3[0]);
        mod.backward(grad);

        predict = mod.forward(x4);
        grad[0] = 2 * (predict[0] - y4[0]);
        mod.backward(grad);

    };
    
    predict = mod.forward(x5);
    std::cout << predict[0];

}

// Запуск программы: CTRL+F5 или меню "Отладка" > "Запуск без отладки"
// Отладка программы: F5 или меню "Отладка" > "Запустить отладку"

// Советы по началу работы 
//   1. В окне обозревателя решений можно добавлять файлы и управлять ими.
//   2. В окне Team Explorer можно подключиться к системе управления версиями.
//   3. В окне "Выходные данные" можно просматривать выходные данные сборки и другие сообщения.
//   4. В окне "Список ошибок" можно просматривать ошибки.
//   5. Последовательно выберите пункты меню "Проект" > "Добавить новый элемент", чтобы создать файлы кода, или "Проект" > "Добавить существующий элемент", чтобы добавить в проект существующие файлы кода.
//   6. Чтобы снова открыть этот проект позже, выберите пункты меню "Файл" > "Открыть" > "Проект" и выберите SLN-файл.

File: nn.py
from random import random

class model():
    def __init__(self):
        pass
    
    def add_layers(self):
        pass
    
    def compile(self):
        pass

    def forward(self):
        pass
    
    def backward(self):
        pass

    def fit(self):
        pass

    def init_w(self):
        pass


class NeuralNetwork(model):

    def __init__(self):
        self.layers = []
        self.lr = 1

    def add_layers(self, layer):
        self.layers.append(layer)

    def init_w(self, size):
        w = []
        for i in range(size):
            w.append(random()-0.5)
        return w
        

    def compile(self):
        self.size = len(self.layers)
        for i in range(1,self.size):
            for neuron in self.layers[i].neurons:
                neuron.w = self.init_w(self.layers[i-1].size)

    def forward(self, X):
        out = []
        for i in range(len(X)):
            self.layers[0].neurons[i].out = X[i]
        for i in range(1, self.size):
            for j in range(self.layers[i].size):
                self.layers[i].neurons[j].sum = 0
                for k in range(self.layers[i-1].size):
                    self.layers[i].neurons[j].sum += self.layers[i-1].neurons[k].out * self.layers[i].neurons[j].w[k] 
                self.layers[i].neurons[j].sum += self.layers[i].neurons[j].b      
            self.layers[i].activate()
        for i in range(self.layers[-1].size):
            out.append(self.layers[-1].neurons[i].out)
        return out
            
    def backward(self, gradient):
        for k in reversed(range(len(self.layers)-1)):
            gradientNext = []
            der = self.layers[k+1].derivative()
            for i in range(self.layers[k+1].size):
                    gradient[i] *= der[i] * self.lr
            for i in range(self.layers[k].size):
                gradientNext.append(0)
                for j in range(self.layers[k+1].size):
                    gradientNext[i] += self.layers[k+1].neurons[j].w[i] * gradient[j]
            
            for i in range(self.layers[k+1].size):
                for j in range(self.layers[k].size):
                    self.layers[k+1].neurons[i].w[j] -= gradient[i] * self.layers[k].neurons[j].out
                self.layers[k+1].neurons[i].b -= gradient[i]

            gradient = gradientNext.copy()



class NN_momentum(NeuralNetwork):
    def __init__(self):
        super().__init__()

    def backward(self, gradient):
        pass
